rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md

```diff
@@ -0,0 +1,394 @@
+have their pros and cons, it would be a benefit if Vector smoothly
+supported the type of behavior users may be accustomed to from those
+other tools.
+
+While users could technically just increase Vector's batch sizes and
+timeouts to approach the desired behavior, the implementation is not
+designed for it and would cause significant problems:
+
+- End-to-end acknowledgements are not issued until the batch is
+  finished, which will cause backpressure to sources and possibly
+  timeouts with retransmissions.
+- The batches are stored in memory which could cause out of memory
+  crashes.
+
+## Proposal
+
+Vector will use the AWS S3 multipart upload API to be able to upload
+smaller batches. Due to the requirements of this API, this will
+require some local storage but will not store full objects.
+
+The AWS S3 multipart upload API uses the following process:
+
+1. The upload is created with the `CreateMultipartUpload` action. This
+   action returns a unique identifier for the upload which is used for
+   all following actions.
+2. Individual parts are uploaded with the `UploadPart` action. These
+   uploads contain a sequence number that identifies the order they
+   will be assembled into the final object, and may be uploaded in any
+   order.
+3. The final object is assembled with the `CompleteMultipartUpload`
+   action.
+
+There are two major issues that this API presents that force some of
+the implementation decisions below:
+
+1. Vector may need to continue a multipart upload process across sink
+   restarts (due to reloads, restarts, or crashes). Since the uploads
+   require the presence of a unique identifier for both the multipart
+   upload itself as well as each of the uploaded parts, Vector will
+   need to store these identifiers with the partition that initiated
+   the upload.
+2. The upload parts (except for the final part) have a _minimum_ size
+   limit of 5MB. Since the sink may need to be restarted before a
+   batch reaches this minimum size, it will also need to buffer the
+   batch to disk before uploading the part.
+
+### User Experience
+
+The enhanced sink will add support for a new time-based upload
+mode. This mode will create a new object in the destination bucket
+once every configured interval instead of creating an object when a
+batch is filled or times out. This new object will be uploaded
+incrementally, but will not be visible in the bucket until all the
+parts of it are complete.
+
+The following upload behaviors are possible through the configuration
+described below:
+
+1. The sink operates as before, where batches are buffered in memory
+   and uploaded directly to objects. This will initially be the
+   default configuration to avoid breaking existing setups.
+1. The sink buffers batches until it can generate a complete part, and
+   assembles the parts into an object only at the end of an
+   interval. None of the intermediate parts are visible until
```

Comment:
Looks like this sentence wasn't finished. And to clarify, this is local batching on Vector's host? Data doesn't reach S3 until the end of the interval?